File: src/routes/remotes.ts
import { Router, type Request, type Response } from 'express';
import { isValidRemoteName, isValidRoutePath, isValidUrlOrPath } from '@bimo-dk/nexus-core';
import {
  addRemote,
  deleteRemote,
  getAllRemotes,
  getRemote,
  RegistryConflictError,
  toggleRemote,
  updateRemote,
} from '../store.js';
import { cid } from '../correlation.js';
import { broadcastRemotesChanged } from '../websocket.js';
import type { AddRemoteRequest, RemoteConfig, UpdateRemoteRequest } from '../types.js';

export const remotesRouter = Router();

remotesRouter.get('/', async (_req: Request, res: Response) => {
  const remotes = await getAllRemotes();
  res.json({
    remotes,
    total: remotes.length,
    enabled: remotes.filter((r) => r.enabled).length,
  });
});

remotesRouter.get('/:name', async (req: Request, res: Response) => {
  const name = String(req.params['name']);
  const remote = await getRemote(name);
  if (!remote) {
    res.status(404).json({ error: 'not_found', message: `Remote "${name}" not found`, correlationId: cid(req) });
    return;
  }
  res.json(remote);
});

remotesRouter.post('/', async (req: Request, res: Response) => {
  const body = req.body as Partial<AddRemoteRequest> | undefined;
  if (!body || typeof body !== 'object') {
    res.status(400).json({ error: 'invalid_body', message: 'JSON body required', correlationId: cid(req) });
    return;
  }

  const { name, url, routePath, upstreamUrl } = body;
  const exposedModule = body.exposedModule ?? './RemoteEntry';
  const enabled = body.enabled ?? true;

  const validationError = validateNewRemote({ name, url, routePath, exposedModule });
  if (validationError) {
    console.error(`[remotes] [${cid(req)}] POST validation failed: ${validationError}`);
    res.status(400).json({ error: 'validation_failed', message: validationError, correlationId: cid(req) });
    return;
  }

  const newRemote: RemoteConfig = {
    name: name!,
    url: url!,
    exposedModule,
    routePath: routePath!,
    enabled,
    addedAt: new Date().toISOString(),
    ...(upstreamUrl ? { upstreamUrl } : {}),
  };

  try {
    const created = await addRemote(newRemote);
    res.status(201).json(created);
    broadcastRemotesChanged(`add:${created.name}`);
  } catch (err) {
    if (err instanceof RegistryConflictError) {
      console.error(`[remotes] [${cid(req)}] POST conflict: ${err.message}`);
      res.status(409).json({ error: 'conflict', message: err.message, correlationId: cid(req) });
      return;
    }
    throw err;
  }
});

remotesRouter.put('/:name', async (req: Request, res: Response) => {
  const name = String(req.params['name']);
  const body = req.body as UpdateRemoteRequest | undefined;
  if (!body || typeof body !== 'object') {
    res.status(400).json({ error: 'invalid_body', message: 'JSON body required', correlationId: cid(req) });
    return;
  }

  if (body.routePath !== undefined && !isValidRoutePath(body.routePath)) {
    res.status(400).json({ error: 'validation_failed', message: 'routePath must be kebab-case starting with a lowercase letter', correlationId: cid(req) });
    return;
  }
  if (body.url !== undefined && !isValidUrlOrPath(body.url)) {
    res.status(400).json({ error: 'validation_failed', message: 'url must be a valid http(s) URL or absolute path', correlationId: cid(req) });
    return;
  }

  const updated = await updateRemote(name, body);
  if (!updated) {
    res.status(404).json({ error: 'not_found', message: `Remote "${name}" not found`, correlationId: cid(req) });
    return;
  }
  res.json(updated);
  broadcastRemotesChanged(`update:${updated.name}`);
});

remotesRouter.delete('/:name', async (req: Request, res: Response) => {
  const name = String(req.params['name']);
  const removed = await deleteRemote(name);
  if (!removed) {
    res.status(404).json({ error: 'not_found', message: `Remote "${name}" not found`, correlationId: cid(req) });
    return;
  }
  res.status(204).send();
  broadcastRemotesChanged(`delete:${name}`);
});

remotesRouter.post('/:name/toggle', async (req: Request, res: Response) => {
  const name = String(req.params['name']);
  const updated = await toggleRemote(name);
  if (!updated) {
    res.status(404).json({ error: 'not_found', message: `Remote "${name}" not found`, correlationId: cid(req) });
    return;
  }
  res.json(updated);
  broadcastRemotesChanged(`toggle:${updated.name}`);
});

remotesRouter.post('/:name/redeploy', async (req: Request, res: Response) => {
  const name = String(req.params['name']);
  const remote = await getRemote(name);
  if (!remote) {
    res.status(404).json({ error: 'not_found', message: `Remote "${name}" not found`, correlationId: cid(req) });
    return;
  }
  console.log(`[registry] [${cid(req)}] Redeploy signal for "${remote.name}" at ${new Date().toISOString()}`);
  res.status(202).json({
    accepted: true,
    remote: remote.name,
    timestamp: new Date().toISOString(),
    correlationId: cid(req),
    note: 'Redeploy is logged. Container orchestration (Docker Swarm/K8s) is responsible for actually redeploying.',
  });
});

function validateNewRemote(input: {
  name: string | undefined;
  url: string | undefined;
  routePath: string | undefined;
  exposedModule: string;
}): string | null {
  if (!input.name || !isValidRemoteName(input.name)) {
    return 'name must be camelCase starting with a lowercase letter';
  }
  if (!input.url || !isValidUrlOrPath(input.url)) {
    return 'url must be a valid http(s) URL or absolute path (starting with /)';
  }
  if (!input.routePath || !isValidRoutePath(input.routePath)) {
    return 'routePath must be kebab-case starting with a lowercase letter';
  }
  if (!input.exposedModule.startsWith('./')) {
    return 'exposedModule must start with "./"';
  }
  return null;
}

File: src/api/remotes.rs
use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::{get, post},
    Extension, Json, Router,
};
use serde::Deserialize;
use serde_json::json;
use tracing::{error, info};

use crate::correlation::CorrelationId;
use crate::http_error::error_response;
use crate::state::AppState;
use crate::store::{self, StoreError};
use crate::time::iso_now;
use crate::types::{AddRemoteRequest, RegistryResponse, RemoteConfig, UpdateRemoteRequest};
use crate::validators::{is_valid_remote_name, is_valid_route_path, is_valid_url_or_path, parse_visibility};
use crate::ws::broadcast_remotes_changed;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", get(list).post(create))
        .route("/{name}", get(detail).put(update).delete(remove))
        .route("/{name}/toggle", post(toggle))
        .route("/{name}/redeploy", post(redeploy))
}

#[derive(Deserialize)]
struct ListQuery {
    #[serde(rename = "host_id")]
    host_id: Option<String>,
}

async fn list(State(state): State<AppState>, Query(q): Query<ListQuery>) -> Response {
    let result = match q.host_id.as_deref() {
        Some(host_id) => store::list_for_host(&state.db, host_id).await,
        None => store::list(&state.db).await,
    };
    match result {
        Ok(remotes) => {
            let enabled = remotes.iter().filter(|r| r.enabled).count();
            let total = remotes.len();
            Json(RegistryResponse {
                remotes,
                total,
                enabled,
            })
            .into_response()
        }
        Err(e) => server_error(e, "<no-correlation-id>"),
    }
}

async fn detail(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Extension(cid): Extension<CorrelationId>,
) -> Response {
    match store::get(&state.db, &name).await {
        Ok(Some(remote)) => Json(remote).into_response(),
        Ok(None) => error_response(
            StatusCode::NOT_FOUND,
            "not_found",
            format!("Remote \"{}\" not found", name),
            cid.as_str(),
        ),
        Err(e) => server_error(e, cid.as_str()),
    }
}

/// Validates the `visibility` string format and (if host-specific) confirms
/// the referenced host actually exists. Returns `Ok(canonical_value)` or `Err(message)`.
async fn validate_visibility(db: &store::Db, raw: &str) -> Result<String, String> {
    match parse_visibility(raw) {
        Ok(None) => Ok("global".to_string()),
        Ok(Some(host_id)) => {
            let exists = store::host_exists(db, host_id)
                .await
                .map_err(|e| format!("database error checking host: {}", e))?;
            if !exists {
                return Err(format!(
                    "visibility references host_id \"{}\" which does not exist",
                    host_id
                ));
            }
            Ok(raw.to_string())
        }
        Err(msg) => Err(msg.to_string()),
    }
}

async fn create(
    State(state): State<AppState>,
    Extension(cid): Extension<CorrelationId>,
    body: Option<Json<AddRemoteRequest>>,
) -> Response {
    let Some(Json(body)) = body else {
        return error_response(
            StatusCode::BAD_REQUEST,
            "invalid_body",
            "JSON body required",
            cid.as_str(),
        );
    };

    let name = body.name.unwrap_or_default();
    let url = body.url.unwrap_or_default();
    let route_path = body.route_path.unwrap_or_default();
    let exposed_module = body.exposed_module.unwrap_or_else(|| "./RemoteEntry".to_string());
    let enabled = body.enabled.unwrap_or(true);
    let raw_visibility = body.visibility.unwrap_or_else(|| "global".to_string());

    if let Some(msg) = validate_new(&name, &url, &route_path, &exposed_module) {
        error!("[remotes] [{}] POST validation failed: {}", cid.as_str(), msg);
        return error_response(StatusCode::BAD_REQUEST, "validation_failed", msg, cid.as_str());
    }
    let visibility = match validate_visibility(&state.db, &raw_visibility).await {
        Ok(v) => v,
        Err(msg) => {
            error!("[remotes] [{}] POST visibility failed: {}", cid.as_str(), msg);
            return error_response(StatusCode::BAD_REQUEST, "validation_failed", msg, cid.as_str());
        }
    };

    let remote = RemoteConfig {
        name,
        url,
        exposed_module,
        route_path,
        enabled,
        added_at: iso_now(),
        upstream_url: body.upstream_url,
        health_status: None,
        last_health_check: None,
        visibility,
    };

    match store::insert(&state.db, &remote).await {
        Ok(()) => {
            let trigger = format!("add:{}", remote.name);
            let res = (StatusCode::CREATED, Json(remote)).into_response();
            broadcast_remotes_changed(&state, trigger).await;
            res
        }
        Err(StoreError::Conflict(name)) => {
            error!(
                "[remotes] [{}] POST conflict: Remote \"{}\" already exists",
                cid.as_str(),
                name
            );
            error_response(
                StatusCode::CONFLICT,
                "conflict",
                format!("Remote \"{}\" already exists", name),
                cid.as_str(),
            )
        }
        Err(e) => server_error(e, cid.as_str()),
    }
}

async fn update(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Extension(cid): Extension<CorrelationId>,
    body: Option<Json<UpdateRemoteRequest>>,
) -> Response {
    let Some(Json(mut body)) = body else {
        return error_response(
            StatusCode::BAD_REQUEST,
            "invalid_body",
            "JSON body required",
            cid.as_str(),
        );
    };

    if let Some(ref rp) = body.route_path {
        if !is_valid_route_path(rp) {
            return error_response(
                StatusCode::BAD_REQUEST,
                "validation_failed",
                "routePath must be kebab-case starting with a lowercase letter",
                cid.as_str(),
            );
        }
    }
    if let Some(ref u) = body.url {
        if !is_valid_url_or_path(u) {
            return error_response(
                StatusCode::BAD_REQUEST,
                "validation_failed",
                "url must be a valid http(s) URL or absolute path",
                cid.as_str(),
            );
        }
    }
    if let Some(ref v) = body.visibility {
        match validate_visibility(&state.db, v).await {
            Ok(canonical) => body.visibility = Some(canonical),
            Err(msg) => {
                return error_response(StatusCode::BAD_REQUEST, "validation_failed", msg, cid.as_str());
            }
        }
    }

    match store::update(&state.db, &name, body).await {
        Ok(Some(remote)) => {
            let trigger = format!("update:{}", remote.name);
            let res = Json(remote).into_response();
            broadcast_remotes_changed(&state, trigger).await;
            res
        }
        Ok(None) => error_response(
            StatusCode::NOT_FOUND,
            "not_found",
            format!("Remote \"{}\" not found", name),
            cid.as_str(),
        ),
        Err(e) => server_error(e, cid.as_str()),
    }
}

async fn remove(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Extension(cid): Extension<CorrelationId>,
) -> Response {
    match store::delete(&state.db, &name).await {
        Ok(true) => {
            let trigger = format!("delete:{}", name);
            let res = StatusCode::NO_CONTENT.into_response();
            broadcast_remotes_changed(&state, trigger).await;
            res
        }
        Ok(false) => error_response(
            StatusCode::NOT_FOUND,
            "not_found",
            format!("Remote \"{}\" not found", name),
            cid.as_str(),
        ),
        Err(e) => server_error(e, cid.as_str()),
    }
}

async fn toggle(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Extension(cid): Extension<CorrelationId>,
) -> Response {
    match store::toggle(&state.db, &name).await {
        Ok(Some(remote)) => {
            let trigger = format!("toggle:{}", remote.name);
            let res = Json(remote).into_response();
            broadcast_remotes_changed(&state, trigger).await;
            res
        }
        Ok(None) => error_response(
            StatusCode::NOT_FOUND,
            "not_found",
            format!("Remote \"{}\" not found", name),
            cid.as_str(),
        ),
        Err(e) => server_error(e, cid.as_str()),
    }
}

async fn redeploy(
    State(state): State<AppState>,
    Path(name): Path<String>,
    Extension(cid): Extension<CorrelationId>,
) -> Response {
    match store::get(&state.db, &name).await {
        Ok(Some(remote)) => {
            let ts = iso_now();
            info!(
                "[registry] [{}] Redeploy signal for \"{}\" at {}",
                cid.as_str(),
                remote.name,
                ts
            );
            (
                StatusCode::ACCEPTED,
                Json(json!({
                    "accepted": true,
                    "remote": remote.name,
                    "timestamp": ts,
                    "correlationId": cid.as_str(),
                    "note": "Redeploy is logged. Container orchestration (Docker Swarm/K8s) is responsible for actually redeploying.",
                })),
            )
                .into_response()
        }
        Ok(None) => error_response(
            StatusCode::NOT_FOUND,
            "not_found",
            format!("Remote \"{}\" not found", name),
            cid.as_str(),
        ),
        Err(e) => server_error(e, cid.as_str()),
    }
}

fn validate_new(name: &str, url: &str, route_path: &str, exposed_module: &str) -> Option<String> {
    if name.is_empty() || !is_valid_remote_name(name) {
        return Some("name must be camelCase starting with a lowercase letter".to_string());
    }
    if url.is_empty() || !is_valid_url_or_path(url) {
        return Some("url must be a valid http(s) URL or absolute path (starting with /)".to_string());
    }
    if route_path.is_empty() || !is_valid_route_path(route_path) {
        return Some("routePath must be kebab-case starting with a lowercase letter".to_string());
    }
    if !exposed_module.starts_with("./") {
        return Some("exposedModule must start with \"./\"".to_string());
    }
    None
}

fn server_error(err: StoreError, cid: &str) -> Response {
    error!("[remotes] [{}] store error: {}", cid, err);
    error_response(
        StatusCode::INTERNAL_SERVER_ERROR,
        "internal_server_error",
        err.to_string(),
        cid,
    )
}
